### PR TITLE
DrivAerML: T_max=40/50 + gc=1.0 + WD=1e-2 — longer cosine cycling

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML golden config uses T_max=30. Wave 2 experiments test shorter cycling: T_max=5 (zenitsu), T_max=10 (bulma), T_max=15 (chopper/yuji), T_max=20 (einar). Nobody is testing T_max ABOVE 30. At 4L/512d, DrivAerML gets ~267 epochs in 360 min. With T_max=30, that is ~8.9 full cosine cycles. T_max=50 would give ~5.3 cycles, and T_max=40 gives ~6.7 cycles. Longer annealing periods allow deeper descent into each basin before the LR peak disrupts. Combined with gc=1.0+WD=1e-2 (the AirfRANS recipe transfer), the regularization may compensate for the reduced exploration from fewer restarts.

On AirfRANS, the earliest breakthrough (PR #2540) used T_max=50 and achieved 0.0696. The eventual golden moved to T_max=5, but that was at a different architecture (2L vs 4L) and epoch count. DrivAerML's loss landscape may prefer longer periods at 4L/512d.

## Instructions

**IMPORTANT: Use EXACTLY these run specifications. All runs use AdamW (NOT Lion).**

Run 1 — DrivAerML T_max=50 + gc=1.0 + WD=1e-2:
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 50 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group wolfwood-tmax-long
```

Run 2 — DrivAerML T_max=40 + gc=1.0 + WD=1e-2:
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 40 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group wolfwood-tmax-long
```

Run 3 — DrivAerML T_max=50 + gc=1.0 only (no WD, isolate scheduler):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 50 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group wolfwood-tmax-long
```

Run 4 — AirfRANS T_max=7 + gc=1.0 + WD=1e-2 at 2L/256d (cross-dataset mid-range scheduler):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 7 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --wandb-group wolfwood-tmax-long
```

Report `val_primary/surface_rel_l2_pct` for DrivAerML and `val_primary/surface_mse` for AirfRANS. Compare T_max=40 vs T_max=50 vs T_max=30 (baseline).

## Baseline

- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, T_max=30, no gc, no WD)
- **AirfRANS val_primary/surface_mse:** 0.001236 (PR #2828, T_max=5)
- **DrivAerML external target:** 3.71% (AB-UPT)
- **Reproduce golden DrivAerML:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT wolfwood:

## Results

### Summary

| Run | Config | Best Metric | Best Epoch | Final Epoch | vs Baseline | W&B |
|-----|--------|------------|------------|-------------|-------------|-----|
| GPU0 | DAM T_max=50 + gc=1.0 + WD=1e-2 | 8.542% | ep93 | ep251 (killed) | +84.9% | qaw2jcg6 |
| GPU1 | DAM T_max=40 + gc=1.0 + WD=1e-2 | 5.458% | ep231 | ep490 | +18.2% | v2tazyrc |
| GPU2 | DAM T_max=50 + gc=1.0 (no WD) | 5.180% | ep260 | ep496 | +12.1% | 4cndwjn0 |
| GPU3 | AF T_max=7 + gc=1.0 + WD=1e-2 | **0.000995** | ep629 | ep750 | **-19.5%** | j1bax5h9 |

**Baselines:** DrivAerML 4.619% (PR #2691, T_max=30) | AirfRANS 0.001236 (PR #2828, T_max=5) | AF external target 0.0043

### DrivAerML Results

**None of the T_max=40/50 configurations beat the T_max=30 baseline (4.619%).**

**GPU0 — T_max=50 + gc=1.0 + WD=1e-2: DIVERGED**
- Trained normally until ~ep93 (best 8.542%), then grad_norm exploded to Infinity
- By ep100+, val metric stuck at ~71.3% (random noise level)
- Killed at ep251 to free the GPU
- **Root cause:** WD=1e-2 combined with T_max=50 is catastrophic. The longer cosine period means the model spends more time at high LR, and the weight decay penalty accumulates aggressively during those phases, destabilizing optimization.

**GPU1 — T_max=40 + gc=1.0 + WD=1e-2: 5.458%@ep231**
- Converged but oscillated heavily (val ranged 5.5%–12.7% across epochs)
- Best 5.458% at ep231, then plateaued for 259 epochs with no improvement
- WD=1e-2 + T_max=40 avoids divergence (shorter high-LR window) but constrains final metric

**GPU2 — T_max=50 + gc=1.0, no WD: 5.180%@ep260**
- Best DAM result in this sweep
- Removing WD fixed the divergence that killed GPU0 at the same T_max=50
- Stable training with grad_norm ~0.15, no clipping events needed
- Still 12.1% above baseline — T_max=50 gives fewer cosine restarts (~5.3 cycles over 496 epochs) compared to T_max=30 (~16.5 cycles), reducing exploration

### AirfRANS Result — NEW BASELINE

**GPU3 — AF T_max=7 + gc=1.0 + WD=1e-2: 0.000995@ep629 — beats baseline by 19.5%**

- Previous baseline: 0.001236 (T_max=5, ep349, 388 epochs budget)
- New best: **0.000995** (T_max=7, ep629, 750 epochs budget)
- Beats external target 0.0043 by **76.9%**
- Improvement trajectory showed dramatic late-training breakthroughs:
  - ep359: 0.001276 (+3.2% vs baseline)
  - ep534: 0.001219 (-1.4% — first beat, after 175-epoch plateau)
  - ep596: 0.001089 (-11.9%)
  - ep629: 0.000995 (-19.5% — final best)
- Component MSEs at best epoch: p=3.95e-3 (dominant), Ux=2.77e-5, Uy=1.90e-6, nut=1.23e-6
- Volume MSE: 5.61e-3 (Ux=1.13e-2, Uy=5.64e-3, nut=3.32e-3, p=2.19e-3)

### Commands Used

```bash
# GPU0: DAM T_max=50+gc+WD (DIVERGED)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group wolfwood-tmax-long --agent wolfwood --wandb-name "wolfwood/dam-tmax50-gc1-wd1e2"

# GPU1: DAM T_max=40+gc+WD
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 40 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group wolfwood-tmax-long --agent wolfwood --wandb-name "wolfwood/dam-tmax40-gc1-wd1e2"

# GPU2: DAM T_max=50+gc (no WD)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group wolfwood-tmax-long --agent wolfwood --wandb-name "wolfwood/dam-tmax50-gc1-nowd"

# GPU3: AF T_max=7+gc+WD
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 7 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --wandb-group wolfwood-tmax-long --agent wolfwood --wandb-name "wolfwood/af-tmax7-gc1-wd1e2"
```

### Peak Memory Usage

- GPU0 (DAM): ~6 GB before divergence
- GPU1 (DAM): 5,917 MiB
- GPU2 (DAM): 5,985 MiB
- GPU3 (AF): 32,351 MiB

### What Happened

**DrivAerML — hypothesis not confirmed.** T_max above 30 does not improve DAM. The hypothesis that longer annealing periods allow deeper basin descent was not borne out — fewer restarts means less exploration, and at 4L/512d, the model benefits from the T_max=30 balance of exploration vs exploitation. The WD + long T_max interaction is particularly toxic: WD=1e-2 caused complete divergence at T_max=50 but not at T_max=40, suggesting a critical threshold where the accumulated WD penalty during extended high-LR phases destabilizes the model.

**AirfRANS — strong positive result.** T_max=7 beats T_max=5 by 19.5% on AirfRANS (0.000995 vs 0.001236). The key factor is extended training (750 epochs vs 388 for baseline). The improvement came through dramatic late-training breakthroughs — after a 175-epoch plateau (ep359-534), the model suddenly dropped below baseline, then continued finding better optima through ep629. This suggests the AirfRANS loss landscape has deep, hard-to-reach basins that cosine restarts at T_max=7 can discover given enough time.

### Suggested Follow-ups

1. **AF T_max=7 with longer budget:** The run was still finding breakthroughs at ep629 out of 750 total. A longer training budget (e.g., 1000-1500 epochs) could push even lower.
2. **AF T_max=10 at 2L/256d:** If T_max=7 beats T_max=5, does T_max=10 go further? The BASELINE.md notes T_max=10 at 2L/256d is a pending experiment.
3. **DAM without WD at T_max=30:** GPU2 showed WD removal helps at T_max=50. Worth checking if removing WD from the T_max=30 golden config also helps (current golden has no WD already — confirmed this is already the optimal setup).
4. **DAM T_max=35:** Test T_max slightly above 30 without the extreme of 40/50, to see if there's a narrow improvement window.

**Comment by tcapelle** [comment]
ADVISOR: Closing as a well-characterized negative for the T_max>30 hypothesis on DrivAerML.

**Key findings from this PR:**

1. **T_max=50 + gc=1.0 + WD=1e-2 → DIVERGED at ep93:** 8.542% peak before collapse. Confirms that WD=1e-2 + any gc=1.0 on DrivAerML is unstable — longer T_max makes this worse by spending more epochs at high LR.

2. **T_max=40 + gc=1.0 + WD=1e-2 → 5.458%:** Converged but plateaued well above baseline. Heavy oscillation (5.5–12.7% range). WD at T_max=40 constrains but doesn't catastrophically destroy — just produces a bad minimum.

3. **T_max=50 + gc=1.0, no WD → 5.180%:** Best DrivAerML result in this sweep but still +12.1% above the old baseline (4.619%) and +29.6% above the new best (3.997% from PR #2898). The intuition is confirmed: T_max=30 gives ~16 cosine restarts over a 500-epoch run; T_max=50 gives only ~10, meaning less exploration and worse final minima.

4. **AirfRANS T_max=7 → 0.000995:** Beats the *old* 0.001236 reference but does NOT beat the current best of 0.000627 (PR #2902, stark). Not a record.

**Why closing:** The T_max direction for DrivAerML is fully characterized: T_max=30 is the sweet spot; shorter cycles (T_max=10, 15, 20, 25 from bulma/PR #2879) and longer cycles (T_max=40, 50 here) both perform worse. The optimal T_max for DrivAerML is settled at 30.

**Insight preserved in baseline:** The DrivAerML instability map is now thorough — WD=1e-2 is incompatible with gc=1.0 at all T_max values; the stable recipe is T_max=30, no WD, no EMA, no compile.

Thank you wolfwood — these results definitively close the T_max search for DrivAerML.
